experimental/oak_baremetal_loader/src/main.rs

```diff
@@ -186,60 +211,64 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
-        bmrng::unbounded_channel::<Vec<u8>, Result<Vec<u8>, oak_idl::Status>>();
+        bmrng::unbounded_channel::<oak_idl::Request, Result<Vec<u8>, oak_idl::Status>>();
 
-    // Spawn task that handles communication with the trusted runtime.
+    // Spawn task to handle communicating with the runtime and receiving responses.
     tokio::spawn(async move {
-        let mut client = {
-            let comms_channel = Box::new(CommsChannel { inner: comms });
-            let client_handler = ClientHandler::new(comms_channel);
-            schema::TrustedRuntimeClient::new(client_handler)
-        };
-
-        let lookup_data =
-            lookup::load_lookup_data(&cli.lookup_data).expect("failed to load lookup data");
-        let encoded_lookup_data =
-            lookup::encode_lookup_data(lookup_data).expect("failed to encode lookup data");
-
-        if let Err(err) = client.update_lookup_data(encoded_lookup_data.into_vec()) {
-            panic!("failed to send lookup data: {:?}", err)
-        }
-
-        let wasm_bytes = fs::read(&cli.wasm)
-            .with_context(|| format!("Couldn't read Wasm file {}", &cli.wasm.display()))
-            .unwrap();
-        let owned_initialization_flatbuffer = {
-            let mut builder = oak_idl::utils::OwnedFlatbufferBuilder::default();
-            let wasm_module = builder.create_vector::<u8>(&wasm_bytes);
-            let initialization_flatbuffer = schema::Initialization::create(
-                &mut builder,
-                &schema::InitializationArgs {
-                    wasm_module: Some(wasm_module),
-                },
-            );
-
-            builder
-                .finish(initialization_flatbuffer)
-                .expect("errored when creating initialization message")
-        };
-        if let Err(err) = client.initialize(owned_initialization_flatbuffer.into_vec()) {
-            panic!("failed to initialize the runtime: {:?}", err)
-        }
-
-        while let Ok((encoded_request, response_dispatcher)) = request_receiver.recv().await {
-            let response = client
-                .handle_user_request(encoded_request)
-                .map(|oak_idl_message| oak_idl_message.into_vec());
+        let mut communication_channel =
+            BaremetalCommunicationChannel::new(Box::new(CommsChannel { inner: comms }));
+        while let Ok((request, response_dispatcher)) = request_receiver.recv().await {
+            // At the moment requests are sent sequentially, and in FIFO order. The next request
+            // is sent only once a response to the previous message has been implemented.
+            // TODO(#2848): Implement message prioritization, and non sequential invocations.
+            let response = communication_channel.invoke(request);
             response_dispatcher.respond(response).unwrap();
         }
     });
 
+    let mut client = schema::TrustedRuntimeAsyncClient::new(Client {
+        request_dispatcher: request_dispatcher.clone(),
```

Comment:
Okay, this may be my C++ background and allergies about constantly copying data around, but just cloning things makes me suspicious. Why can't it be a mutable reference or take ownership?

Reply:
At the current stage of the code it could be a mutable reference (if we'd modify the client struct to take a lifetime param). However, sooner or later we'll need multiple senders, and might well want to use this client at the same time as others.

Cloning suspicions in general are imo not misplaced, but it's fine here. The ability to clone senders to create multiple of them is pretty much the point of mspc tokio channels, see: https://docs.rs/tokio/latest/tokio/sync/mpsc/fn.channel.html (bmrng is a thin abstraction top of that)

Reply:
Under the hood the sender half uses atomic reference counting, so cloning should be extremely cheap in practice. 

https://github.com/tokio-rs/tokio/blob/4daeea8cad1ce8e67946bc0e17d499ab304b5ca2/tokio/src/sync/mpsc/chan.rs